### DISCUSSION TITLE
SH-RPi 2.0 raspberry shutdown

### DISCUSSION THREAD

**Discussion by psm12** [hatlabs/discussions#59] [Q&A]
Using 'shrpi shutdown' shuts down my pi4 but then power is reapplied by the sh a few seconds later and it reboots. 

Using 'sudo shutdown now' shuts down the pi4 (I think) but the hdmi output continues to show 'Plymouth-power off.service'

Is this expected behaviour? 

Also, is the only way to wake it up to power cycle the whole system ?
Thanks
Paul
Additional Information:
o/s openplotter 3
Sh f/w 2.0.3
Daemon 2.2.2

**Comment by mairas** [comment]
Yes, restarting is the intended behavior. There is no way to know if the shutdown was initiated by a power cut and the power was then subsequently restored or if the shutdown was intended. If you want to shutdown the device while keeping the power on, you can do that with `shrpi sleep` and giving it a timestamp far in the future.
.
At the moment there is no way to trigger a wakeup but to wait until the timeout has lapsed. That could be fixed with minor hardware and firmware hacking. I can provide details if you're interested.

**Reply by psm12** [reply]
Thanks Matti,

I would like to  be able to trigger a wake-up but it certainly isn’t an urgent requirement, it just seems more elegant than a full power cycle  - although that is what is needed with a bare pi4.

I see there is an input to the microcontroller pin accessible on a header that might be useful for this purpose?

Paul

**Reply by psm12** [reply]
Matti,

I'm just tinkering with the 'shrpi sleep' command . It works as I expected using a number (seconds) argument but if I use a time format (00:00 or 00:00:00), I get the response:
"Error: Received HTTP status 400"

(Edit: maybe because I don't have a battery in the sh rtc?)

Also, thinking about the 'shrpi shutdown' command, why does the response to the Linux 'shutdown' command not trigger the watchdog to reboot the pi?  

I'm also wondering what the difference is between 'shrpi sleep' and shrpi shutdown'. Maybe I'm overthinking? 

Paul

**Reply by psm12** [reply]
Hi Matti,

I'm still trying to diagnose why my raspberry pi 4 is not shutting down correctly. Hardware setup pi4, sh2.0, waveshare dual can I/f.

I installed a fresh copy of Openplotter 3 64bit (2023-08-04). Following this shutdown -h now works as expected (plymouth.shutdown.sys displayed then pi powered off). I then used the command string in the getting started document in your doc to install sh and the waveshare can apps. Rebooted and then the shutdown -h now command gave the plymouth.shutdown.sys display but the pi did not shut down.

Paul

**Reply by psm12** [reply]
I have now eliminated Openplotter by repeating the test with the official 64bit (with desktop) raspberry pi o/s with the same result as above.
Paul

**Reply by mairas** [reply]
Hi, I have mostly tested and used the devices in a headless mode (no display or keyboard/mouse). I now connected the Pi to a KVM, and to my amazement, I'm getting the same symptoms as you. The Pi won't shut down when either the USB or the HDMI cable is attached and only shuts down correctly when both are plugged off. This is of course incorrect; I'll investigate.

**Reply by mairas** [reply]
So, as a background, the SH-RPi install script adds the following line to the Pi `/boot/config.txt` configuration file:

```
dtoverlay=gpio-poweroff,gpiopin=2,input,active_low=17
```

This modifies the Pi shutdown process so that instead of rebooting itself and letting the Pi firmware decide whether it should reboot or play dead (that is, remain shut down), it pulls GPIO pin 2 low to indicate the SH-RPi firmware that the Pi has reached poweroff and it's OK to cut the power. 

For some reason, this no longer plays nice with the plymouth thingy that provides the graphical bootup and hides the command line boot messages. I first tried disabling plymouth to see what's going on by removing `quiet splash plymouth.ignore-serial-consoles` from `/boot/cmdline.txt`. Now, the boot messages are shown but the Pi also shuts down and reboots properly! This is of course not a proper fix, so I'll keep investigating.

**Reply by mairas** [reply]
I believe I have found the culprit. On my test device, I had previously had the CAN and RS485 HATs installed and they were also enabled in the software. After removing the hardware, the kernel was all the time complaining about spurious interrupts and wouldn't complete powering off. The correct fix is to run the installer again and disable the interfaces you don't actually have. Alternatively, you can edit `/boot/config.txt` and comment out the `mcp2515-can0` and `sc16is752-spi1` dtoverlay lines towards the end of the file.

Once there reboot once by pulling the power and waiting for the supercaps to deplete. After that, shutting down and rebooting should work as expected.

**Reply by psm12** [reply]
Matti,

I have only the can hat installed in addition to the sailor hat, I've always disabled the other hats in the install dialogues and have also tried it with the can hat enabled and disabled but I always had the shutdown problem. 

Paul

**Reply by mairas** [reply]
Could you share the bottom part (the [all] section) of your `/boot/config.txt` file?

**Reply by psm12** [reply]
[config.txt](https://github.com/hatlabs/discussions/files/12446687/config.txt)

**Reply by mairas** [reply]
OK, that looks completely benign to me. But could we take a step back here - exactly what functionality is not working?

- If the Pi is on and you cut the power, what happens? 
- If the Pi is on and you try to run `sudo shutdown -r now` or `sudo reboot`, what happens?
- If the Pi is on and you try to run `shrpi sleep 120`, what happens?

If you remove `quiet splash plymouth.ignore-serial-consoles` from `/boot/cmdline.txt`, and reboot and then try to do whatever doesn't work, what do you get on the console?

**Comment by mairas** [comment]
If you want to create a custom reboot or reset button, follow these steps:

On the SH-RPi board, close the GPIO4 solder jumper on the bottom side of the board by applying a blob of solder on it. This connects a GPIO pin on the microcontroller to Raspberry Pi's GPIO pin 4. You don't need to provide the signal to the Pi, but this way, you can use Raspberry Pi's 40-pin header to connect a button.

Next, you need to edit the firmware at https://github.com/hatlabs/SH-RPi-firmware. In `src/main.cpp`, there is a commented-out section around line 183. There, uncomment one of the blocks and replace the pin with `PIN_PB5`. Or implement some custom functionality. :-)

**Comment by psm12** [comment]
Matti,

Here are the results:

exactly what functionality is not working?
If the Pi is on and you cut the power, what happens?

Running lights on sailor hat
Plymouth-power off.service very briefly appears.  

Everything powers off quickly

-powered back on

Three leds lit on the sailor hat and pi then boots quickly

If the Pi is on and you try to run sudo shutdown -r now 

Plymouth-power off.service displayed continuously. 
Pi4(8gb) activity and power light both off
All four sailor hat leds on continuously
powered off
Sailor hat shuts down immediately 
powered back on
Sailor hat and pi boot immediately 


or sudo reboot, what happens?

A successful reboot aster Plymouth-reboot.system displayed for a few seconds

If the Pi is on and you try to run shrpi sleep 120, what happens?

Plymouth-shutdown.service displayed for a second

Pi leds go out

Sailor hat leds flash every~ 2 seconds

Pi reboots after delay

I hope this is helpful. 
Paul

**Reply by mairas** [reply]
Everything else seems normal, but `sudo shutdown -r now` and `sudo reboot` should have identical behavior. (Apparently there may be some minor differences in the reboot process, but they are not significant in this case.) Are you sure you gave that exact command?

**Reply by psm12** [reply]
"Everything else seems normal, but sudo shutdown -r now and sudo reboot should have identical behavior. (Apparently there may be some minor differences in the reboot process, but they are not significant in this case.) Are you sure you gave that exact command?"

Yes, exactly. 

Paul

**Reply by psm12** [reply]
![image](https://github.com/hatlabs/discussions/assets/84899243/1b36aa8a-de12-4630-97c2-4faa484ada7c)

**Reply by psm12** [reply]
https://github.com/hatlabs/discussions/assets/84899243/ab587ae0-16bc-404f-bb8d-6b398c2e8f94

**Reply by mairas** [reply]
The command you were running in the video was `shutdown -h now`, and it works exactly as expected. I hope you would've told a bit earlier that was what you were attempting.

See my very first reply in this discussion: https://github.com/hatlabs/discussions/discussions/59#discussioncomment-6731640. As I said already there, a bare shutdown command does not work. Without running `shrpi shutdown`, the MCU is never informed about the shutdown and doesn't know to cut the power. That results in the kernel panicking because it's still alive against its own expectations.

It would probably be possible to make a regular shutdown work by creating another system service that would be triggered on a shutdown (before shrpid is closed). That service would then call `shrpid shutdown` to inform the MCU on the shutdown. But I really don't see the point; the Pi would restart after 5 seconds because the power input wasn't cut, and there's no easy way around that.

**Reply by psm12** [reply]
Matti,

I think as well as confusing you I've also confused myself. 

I had naively thought the following would occur:

Shrpi sleep - as the current function (ie watchdog reboot action inhibited until Wake-up)

Shrpi shutdown - orderly shutdown of pi and sh in some 'standby' mode (watchdog action inhibited)

Shutdown -h now ( or any Linux equivalent eg poweroff or the gui shutdown dialogue or any operating system hang etc.) causes the watchdog to fire and the pi to  be power cycled by the sailor hat to reboot it. I'm very confused about the operation of the sailor hat watchdog. 

Are you saying that the Linux shutdown -h or any equivalent should not be used with the sailor hat attached and although intuitively you would expect shrpi shutdown to shut the pi down, it can only be used to reboot it? 

Apologies i am not a programmer and im not that familiar with Linux. 

Paul

**Reply by psm12** [reply]
Hi Matti,

I'm wondering if in some of your responses you typed shrpi and the spellchecker changed this to sudo? 

Paul

**Reply by mairas** [reply]
The watchdog gets deactivated whenever you stop the shrpid process (e.g. by `sudo systemctl stop shrpid`). And that gets done automatically whenever the system shuts down. That's why the device never reboots if you execute a manual shutdown. If the watchdog weren't deactivated, it would trigger a watchdog reboot e.g. when the Pi was given a reboot command.

`shrpi shutdown` informs the MCU that the Pi is shutting down. The MCU then knows to watch the GPIO pin 2 - when it gets pulled down for a sufficiently long time, it can tell the Pi has shut down, and cuts the power. But as long as the input voltage is present, it _has_ to restart the Pi.

Yes, you're right that a software-initiated shutdown really isn't a thing when using the SH-RPi. It is meant to keep the Pi running at all costs whenever input voltage is present, and to shut it down in a controlled fashion otherwise.  Instead of trying to run shutdown manually and trying to add a button for rebooting, just add a switch to the input voltage wire.

**Reply by psm12** [reply]
I must admit I had expected the opposite watchdog behaviour, so absolutely anything that disrupted the shrpi daemon would cause the watchdog to trigger and reboot the system. What events would cause the watchdog to trigger?
Paul

**Reply by mairas** [reply]
The watchdog is triggered if the watchdog is active and the MCU no longer receives messages from the daemon. For example, if the daemon or the Pi kernel crashes, or if there's a hardware crash on the Pi. If the daemon is stopped in a controlled fashion, it sends a message to disable the watchdog before exiting.

You can test this by shorting SCL and GND on the 40-pin header with a jumper. After 10 seconds, you'll get a bit of dancing LEDs and the Pi will be power-cycled.

**Reply by psm12** [reply]
Thanks Matti,

Sorry to take up your time especially at the weekend. 
Thanks for increasing my understanding. 
The devil is in the detail. 

:) Paul

**Comment by psm12** [comment]
Hi Matti,

Lots on the console, scrolls way over the size of my screen. 

There were a few warnings after all the ok lines. 

Paul
